File: src/main/java/org/panda_lang/nanomaven/console/StatusCommand.java
/*
 * Copyright (c) 2020 Dzikoysk
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.panda_lang.nanomaven.console;

import org.panda_lang.nanomaven.NanoConstants;
import org.panda_lang.nanomaven.NanoMaven;
import org.panda_lang.utilities.commons.console.Effect;

final class StatusCommand implements NanoCommand {

    @Override
    public boolean call(NanoMaven nanoMaven) {
        NanoMaven.getLogger().info("");
        NanoMaven.getLogger().info("NanoMaven " + NanoConstants.VERSION + " Status");
        NanoMaven.getLogger().info("  Active: " + Effect.GREEN_BOLD + nanoMaven.getHttpServer().isAlive() + Effect.RESET);
        NanoMaven.getLogger().info("  Uptime: " + format(nanoMaven.getUptime() / 1000.0 / 60.0) + "min");
        NanoMaven.getLogger().info("  Memory usage of process: " + format((Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()) / 1024.0D / 1024.0D) + "M");
        NanoMaven.getLogger().info("  Cached elements: " + nanoMaven.getMetadataService().getCacheSize());
        nanoMaven.getHttpServer().getLatestError().peek(throwable -> NanoMaven.getLogger().error(" Latest exception", throwable));
        NanoMaven.getLogger().info("");

        return true;
    }

    private String format(double number) {
        return String.format("%.2f", number);
    }

}

File: reposilite-backend/src/main/kotlin/com/reposilite/console/StatusCommand.kt
/*
 * Copyright (c) 2021 dzikoysk
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.reposilite.console

import com.reposilite.Reposilite
import com.reposilite.VERSION
import com.reposilite.console.Status.SUCCEEDED
import com.reposilite.shared.TimeUtils
import com.reposilite.shared.TimeUtils.getPrettyUptimeInMinutes
import panda.utilities.IOUtils
import panda.utilities.console.Effect.GREEN
import panda.utilities.console.Effect.GREEN_BOLD
import panda.utilities.console.Effect.RED_UNDERLINED
import panda.utilities.console.Effect.RESET
import picocli.CommandLine.Command

@Command(name = "status", description = ["Display summary status of app health"])
internal class StatusCommand(
    private val reposilite: Reposilite,
    private val remoteVersionUrl: String,
) : ReposiliteCommand {

    override fun execute(output: MutableList<String>): Status {
        val latestVersion =
            if (reposilite.parameters.testEnv) VERSION
            else getVersion()

        output.add("Reposilite $VERSION Status")
        output.add("  Active: $GREEN_BOLD${reposilite.webServer.isAlive()}$RESET")
        output.add("  Uptime: ${getPrettyUptimeInMinutes(reposilite.startTime)}")
        output.add("  Memory usage of process: ${memoryUsage()}")
        output.add("  Active threads in group: ${threadGroupUsage()}")
        output.add("  Recorded failures: ${reposilite.failureFacade.getFailures().size}")
        output.add("  Latest version of Reposilite: $latestVersion")

        return SUCCEEDED
    }

    private fun getVersion(): String =
        IOUtils.fetchContent(remoteVersionUrl)
            .orElseGet { "$remoteVersionUrl is unavailable: ${it.message}" }
            .let { (if (VERSION == it) GREEN else RED_UNDERLINED).toString() + it + RESET }

    private fun memoryUsage(): String =
        TimeUtils.format((Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()) / 1024.0 / 1024.0) + "M"

    private fun threadGroupUsage(): String =
        Thread.activeCount().toString()

}